src/openfermion/ops/_quadratic_hamiltonian.py

```diff
@@ -253,36 +254,108 @@ def diagonalizing_bogoliubov_transform(self):
            \\end{pmatrix},
 
         where :math:`W` is a :math:`2N \\times 2N` unitary matrix.
+        However, if the Hamiltonian conserves particle number then
+        creation operators don't need to be mixed with annihilation operators
+        and :math:`W` only needs to be an :math:`N \\times N` matrix:
+
+        .. math::
+
+           \\begin{pmatrix}
+                b^\dagger_1 \\\\
+                \\vdots \\\\
+                b^\dagger_N \\\\
+           \\end{pmatrix}
+           = W
+           \\begin{pmatrix}
+                a^\dagger_1 \\\\
+                \\vdots \\\\
+                a^\dagger_N \\\\
+           \\end{pmatrix},
+
         This method returns the matrix :math:`W`.
 
         Returns:
             diagonalizing_unitary (ndarray):
-                A (2 * `n_qubits`) x (2 * `n_qubits`) matrix representing
-                the transformation :math:`W` of the fermionic ladder operators.
+                A matrix representing the transformation :math:`W` of the
+                fermionic ladder operators. If the Hamiltonian conserves
+                particle number then this is :math:`N \\times N`; otherwise
+                it is :math:`2N \\times 2N`.
         """
-        majorana_matrix, majorana_constant = self.majorana_form()
+        if self.conserves_particle_number:
+            energies, diagonalizing_unitary_T = numpy.linalg.eigh(
+                    self.combined_hermitian_part)
+            return diagonalizing_unitary_T.T
+        else:
+            majorana_matrix, majorana_constant = self.majorana_form()
+
+            # Get the orthogonal transformation that puts majorana_matrix
+            # into canonical form
+            canonical, orthogonal = antisymmetric_canonical_form(
+                    majorana_matrix)
+
+            # Create the matrix that converts between fermionic ladder and
+            # Majorana bases
+            normalized_identity = (numpy.eye(self.n_qubits, dtype=complex) /
+                                   numpy.sqrt(2.))
+            majorana_basis_change = numpy.eye(
+                2 * self.n_qubits, dtype=complex) / numpy.sqrt(2.)
+            majorana_basis_change[self.n_qubits:, self.n_qubits:] *= -1.j
+            majorana_basis_change[:self.n_qubits,
+                                  self.n_qubits:] = normalized_identity
+            majorana_basis_change[self.n_qubits:,
+                                  :self.n_qubits] = 1.j * normalized_identity
 
-        # Get the orthogonal transformation that puts majorana_matrix
-        # into canonical form
-        canonical, orthogonal = antisymmetric_canonical_form(majorana_matrix)
+            # Compute the unitary and return
+            diagonalizing_unitary = majorana_basis_change.T.conj().dot(
+                orthogonal.dot(majorana_basis_change))
+
+            return diagonalizing_unitary
+
+    def diagonalizing_circuit(self):
```

Comment:
I'm somewhat thinking that this should actually be a function that is called on the class as opposed to a method. One could imagine wanting to use some of this functionality without initializing the class. Also, the single responsibility principle is a general design pattern that we try to follow.

Reply:
I just tried to be consistent with what we already have, which seems to be something like "methods are okay if they are really specific to the class and can't be called on other classes." For instance, SymbolicOperator, BinaryPolynomial, and PolynomialTensor all have their own methods, and QuadraticHamiltonian already has methods like `majorana_form` and `diagonalizing_bogoliubov_transform` which I think should be treated the same way.

Reply:
Also, the new method `diagonalizing_circuit` depends on the method `diagonalizing_bogoliubov_transform`, which in the case of non-particle-number-conservation depends on `majorana_form`, so it seemed to make most sense to require a QuadraticHamiltonian to be instantiated, especially if we want to have one function that works for both particle-number-conserving and not.